### DISCUSSION TITLE
How does the `CreateCommandsInvoker` work?

### DISCUSSION THREAD

**Discussion by multimeric** [ytausch/serverless-discord-bot#35] [Q&A] [answered]
I gather that the `CreateCommandsFunction` is called once per command deploy, but I can't work out *how* that happens. The `CreateCommandsInvoker` depends on both the `CreateCommandsFunction` and `CommandsLayer`, which makes sense, but what actually *calls* the the `CreateCommandsFunction`?

**Comment by multimeric** [comment]
The reason I ask is because I want to add some on-setup actions (populating a database), but I'm not sure how to ensure that it runs for each deployment.

**Comment by ytausch** [comment, marked as answer]
Hi,

`CreateCommandsFunction` is the lambda backend of an [AWS Lambda-backed custom resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html). In fact, `CreateCommandsInvoker` **is** exactly the name of the custom resource which declares `CreateCommandsFunction` as its backend. You can see that by looking at the `Type` attribute of `CreateCommandsInvoker` in `template.yaml` - it has the format `Custom::XXX`.

That means: If `CreateCommandsInvoker` is created during stack creation, CloudFormation itself [makes a request](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-requests.html) to `CreateCommandsFunction` which is then responsible for setting up the commands and responding back to CloudFormation that everything went fine. A lot of this request and response logic happens in our dependency [cfn-lambda](https://github.com/andrew-templeton/cfn-lambda).

The same thing happens when any commands are updated, it is just a little bit more complicated: Updating the commands means updating the commands layer, which then means that the version number of the commands layer is incremented. Since the ARN of the commands layer is passed as property to `CreateCommandsInvoker` and the ARN of a lambda layer ends with the layer version number, the properties of `CreateCommandsInvoker` change, which triggers a resource update. `CreateCommandsFunction` is then again called to handle the update.

If you want to implement a custom setup logic, you have essentially two options:

1. Extend `CreateCommandsFunction` (note that you can differentiate between creation and update events)
2. Create a new custom lambda-backed stack resource similar to `CreateCommandsInvoker` that has its own lambda handler similar to `CreateCommandsFunction` (I recommend also using cfn-lambda there for doing the heavy-lifting).
